File: apps/services/notifications/Notifications.Infrastructure/Webhooks/Verifiers/SendGridVerifier.cs
using System.Security.Cryptography;
using System.Text;
using Microsoft.Extensions.Logging;

namespace Notifications.Infrastructure.Webhooks.Verifiers;

public class SendGridVerifier
{
    private readonly bool _enabled;
    private readonly string _publicKey;
    private readonly string _environment;
    private readonly ILogger<SendGridVerifier> _logger;

    public SendGridVerifier(bool enabled, string publicKey, string environment, ILogger<SendGridVerifier> logger)
    {
        _enabled = enabled;
        _publicKey = publicKey;
        _environment = environment;
        _logger = logger;
    }

    public (bool Verified, bool Skipped, string? Reason) Verify(string rawBody, string? signature, string? timestamp)
    {
        if (!_enabled)
        {
            _logger.LogWarning(
                "SendGrid webhook verification is disabled; request rejected. " +
                "Set SENDGRID_WEBHOOK_VERIFICATION_ENABLED=true and configure " +
                "SENDGRID_WEBHOOK_PUBLIC_KEY to accept webhook callbacks.");
            return (false, false, "verification_disabled");
        }

        if (string.IsNullOrEmpty(_publicKey))
        {
            _logger.LogError("SendGrid webhook verification is enabled but no public key is configured");
            return (false, false, "missing_public_key");
        }

        if (string.IsNullOrEmpty(signature) || string.IsNullOrEmpty(timestamp))
        {
            _logger.LogWarning("SendGrid webhook request missing signature or timestamp headers");
            return (false, false, "missing_headers");
        }

        try
        {
            var payload = Encoding.UTF8.GetBytes(timestamp + rawBody);
            var publicKeyBytes = Convert.FromBase64String(_publicKey);

            using var ecdsa = ECDsa.Create();
            ecdsa.ImportSubjectPublicKeyInfo(publicKeyBytes, out _);

            var signatureBytes = Convert.FromBase64String(signature);
            var valid = ecdsa.VerifyData(payload, signatureBytes, HashAlgorithmName.SHA256);

            if (!valid)
            {
                _logger.LogWarning("SendGrid webhook signature verification failed");
                return (false, false, "invalid_signature");
            }

            return (true, false, null);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "SendGrid webhook verification threw an error");
            return (false, false, $"verification_error: {ex.Message}");
        }
    }
}

File: _archived/notifications-nodejs/src/integrations/webhooks/verifiers/sendgrid.verifier.ts
import * as crypto from "crypto";
import { logger } from "../../../shared/logger";

interface SendGridVerifierConfig {
  enabled: boolean;
  publicKey: string;
  nodeEnv: string;
}

export interface VerificationResult {
  verified: boolean;
  skipped: boolean;
  reason?: string;
}

/**
 * Verifies SendGrid Event Webhook signatures using ECDSA (P-256 + SHA256).
 *
 * Headers used:
 *   X-Twilio-Email-Event-Webhook-Signature — base64 ECDSA signature
 *   X-Twilio-Email-Event-Webhook-Timestamp — Unix timestamp string
 *
 * Verification payload: timestamp + rawBody (no separator)
 */
export class SendGridWebhookVerifier {
  private config: SendGridVerifierConfig;

  constructor(config: SendGridVerifierConfig) {
    this.config = config;
  }

  verify(
    rawBody: string,
    signature: string | undefined,
    timestamp: string | undefined
  ): VerificationResult {
    const { enabled, publicKey, nodeEnv } = this.config;

    if (!enabled) {
      if (nodeEnv === "production") {
        logger.warn(
          "SendGrid webhook verification is disabled in production — requests are accepted without signature check"
        );
      }
      return { verified: false, skipped: true, reason: "verification_disabled" };
    }

    if (!publicKey) {
      logger.error("SendGrid webhook verification is enabled but no public key is configured");
      return { verified: false, skipped: false, reason: "missing_public_key" };
    }

    if (!signature || !timestamp) {
      logger.warn("SendGrid webhook request missing signature or timestamp headers");
      return { verified: false, skipped: false, reason: "missing_headers" };
    }

    try {
      const payload = timestamp + rawBody;
      const ecPublicKey = crypto.createPublicKey({
        key: Buffer.from(publicKey, "base64"),
        format: "der",
        type: "spki",
      });

      const verifier = crypto.createVerify("SHA256");
      verifier.update(payload, "utf8");
      const valid = verifier.verify(ecPublicKey, signature, "base64");

      if (!valid) {
        logger.warn("SendGrid webhook signature verification failed");
        return { verified: false, skipped: false, reason: "invalid_signature" };
      }

      return { verified: true, skipped: false };
    } catch (err) {
      logger.error("SendGrid webhook verification threw an error", { error: String(err) });
      return { verified: false, skipped: false, reason: `verification_error: ${String(err)}` };
    }
  }
}
